File: src/services/storyStore.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Story } from '@/models/types';

/**
 * On-device persistence + pure helpers for the story collection.
 *
 * AsyncStorage works across iOS, Android and web (web uses localStorage), so
 * everything stays on-device with no accounts or cloud sync.
 */

const STORAGE_KEY = 'storybrew.stories.v1';

export async function loadStories(): Promise<Story[]> {
  try {
    const raw = await AsyncStorage.getItem(STORAGE_KEY);
    if (!raw) return [];
    const parsed = JSON.parse(raw);
    return Array.isArray(parsed) ? (parsed as Story[]) : [];
  } catch {
    return [];
  }
}

export async function saveStories(stories: Story[]): Promise<void> {
  try {
    await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(stories));
  } catch {
    // Best-effort; ignore write failures.
  }
}

// ---- Pure selectors (no side effects) ----

/** The playlist: approved stories, newest first. */
export function selectPlaylist(stories: Story[]): Story[] {
  return stories
    .filter((s) => s.approvalState === 'approved')
    .sort((a, b) => b.createdAt - a.createdAt);
}

/** Approved + pending episodes that belong to the same series, in order. */
export function selectEpisodesInSeriesOf(stories: Story[], story: Story): Story[] {
  if (!story.seriesId) return [story];
  return stories
    .filter((s) => s.seriesId === story.seriesId)
    .sort((a, b) => a.episodeNumber - b.episodeNumber);
}

/** Summaries of approved earlier episodes, to keep a new episode consistent. */
export function selectPreviousSummaries(stories: Story[], seriesId: string): string[] {
  return stories
    .filter((s) => s.seriesId === seriesId && s.approvalState === 'approved')
    .sort((a, b) => a.episodeNumber - b.episodeNumber)
    .map((s) => s.summary);
}

export function nextEpisodeNumber(stories: Story[], seriesId: string): number {
  const highest = stories
    .filter((s) => s.seriesId === seriesId)
    .reduce((max, s) => Math.max(max, s.episodeNumber), 0);
  return highest + 1;
}

File: StoryBrew/Services/StoryStore.swift
import Foundation
import Combine

/// The single source of truth for saved stories. Persists to a JSON file in the
/// app's Documents directory, so everything stays on-device. Drives the
/// playlist (approved stories) and supports deletion and series grouping.
@MainActor
final class StoryStore: ObservableObject {
    @Published private(set) var stories: [Story] = []

    private let fileURL: URL

    init(filename: String = "stories.json") {
        let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        self.fileURL = documents.appendingPathComponent(filename)
        load()
    }

    // MARK: - Derived collections

    /// The playlist: approved stories, newest first.
    var playlist: [Story] {
        stories
            .filter { $0.approvalState == .approved }
            .sorted { $0.createdAt > $1.createdAt }
    }

    var hasPlaylist: Bool { !playlist.isEmpty }

    /// Approved episodes that belong to the same series as `story`, in order.
    func episodes(inSeriesOf story: Story) -> [Story] {
        guard let seriesID = story.seriesID else { return [story] }
        return stories
            .filter { $0.seriesID == seriesID }
            .sorted { $0.episodeNumber < $1.episodeNumber }
    }

    /// Summaries of approved earlier episodes, used to keep a new episode consistent.
    func previousSummaries(seriesID: UUID) -> [String] {
        stories
            .filter { $0.seriesID == seriesID && $0.approvalState == .approved }
            .sorted { $0.episodeNumber < $1.episodeNumber }
            .map(\.summary)
    }

    func nextEpisodeNumber(seriesID: UUID) -> Int {
        let highest = stories.filter { $0.seriesID == seriesID }.map(\.episodeNumber).max() ?? 0
        return highest + 1
    }

    // MARK: - Mutations

    func upsert(_ story: Story) {
        if let index = stories.firstIndex(where: { $0.id == story.id }) {
            stories[index] = story
        } else {
            stories.append(story)
        }
        save()
    }

    func approve(_ story: Story) {
        update(story.id) { $0.approvalState = .approved }
    }

    func reject(_ story: Story) {
        update(story.id) { $0.approvalState = .rejected }
    }

    func markAudioReady(_ storyID: UUID) {
        update(storyID) { $0.audioReady = true }
    }

    func delete(_ story: Story) {
        stories.removeAll { $0.id == story.id }
        save()
    }

    func delete(at offsets: IndexSet, in list: [Story]) {
        let ids = offsets.map { list[$0].id }
        stories.removeAll { ids.contains($0.id) }
        save()
    }

    private func update(_ id: UUID, _ change: (inout Story) -> Void) {
        guard let index = stories.firstIndex(where: { $0.id == id }) else { return }
        change(&stories[index])
        save()
    }

    // MARK: - Persistence

    private func load() {
        guard let data = try? Data(contentsOf: fileURL) else { return }
        if let decoded = try? JSONDecoder.storyDecoder.decode([Story].self, from: data) {
            stories = decoded
        }
    }

    private func save() {
        guard let data = try? JSONEncoder.storyEncoder.encode(stories) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }
}

private extension JSONEncoder {
    static var storyEncoder: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }
}

private extension JSONDecoder {
    static var storyDecoder: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }
}
